RULES.md

```diff
@@ -13,243 +13,474 @@ Notices are split into three categories: `INFO`, `WARNING`, `ERROR`.
 
-<a name="MissingRequiredColumn"/>
+The combination of `fare_rules.route_id`, `fare_rules.origin_id`, `fare_rules.contains_id` and `fare_rules.destination_id` fields should be unique in GTFS file `fare_rules.txt`.
 
-### MoreThanOneEntityNotice
+#### References:
+* [Original Python validator implementation](https://github.com/google/transitfeed)
 
-The file is expected to have a single entity but has more (e.g., "feed_info.txt").
+<a name="DuplicateKeyError"/>
 
-<a name="MissingRequiredFileError"/>
+### DuplicateKeyError
 
-### MissingRequiredFileError
+The values of the given key and rows are duplicates.
 
-A required file is missing.
+#### References:
+* [Original Python validator implementation](https://github.com/google/transitfeed)
+
+<a name="EmptyFileNotice"/>
+
+### EmptyFileNotice
+
+Empty csv file found in the archive: file does not have any headers, or is a required file and does not have any data. The GTFS specification requires the first line of each file to contain field names and required files must have data.
+
+#### References:
+* [GTFS files requirements](https://gtfs.org/reference/static#file-requirements)
+
+<a name="ForeignKeyError"/>
+
+### ForeignKeyError
+
+The values of the given key and rows of one table cannot be found a values of the given key in another table.
+
+#### References:
+* [Original Python validator implementation](https://github.com/google/transitfeed)
+
+<a name="InconsistentAgencyTimezoneNotice"/>
+
+### InconsistentAgencyTimezoneNotice
+
+Agencies from GTFS `agency.txt` have been found to have different timezones.
+
+#### References:
+* [GTFS agency.txt specification](https://gtfs.org/reference/static/#agencytxt)
+
+<a name="IOError"/>
+
+### IOError
+
+Error in IO operation.
+
+<a name="InvalidColorNotice"/>
+
+### InvalidColorNotice
+
+Value of field with type `color` is not valid. A color must be encoded as a six-digit hexadecimal number. The leading "#" is not included.
+
+#### References:
+* [Field Types Description](http://gtfs.org/reference/static/#field-types)
+
+<a name="InvalidCurrencyNotice"/>
+
+### InvalidCurrencyNotice
+
+Value of field with type `currency` is not valid. Currency code must follow <a href="https://en.wikipedia.org/wiki/ISO_4217#Active_codes">ISO 4217</a>
+
+#### References:
+* [Field Types Description](http://gtfs.org/reference/static/#field-types)
+
+<a name="InvalidDateNotice"/>
+
+### InvalidDateNotice
+
+Value of field with type `date` is not valid. Dates must have the YYYYMMDD format.
+
+#### References:
+* [Field Types Description](http://gtfs.org/reference/static/#field-types)
+
+<a name="InvalidEmailNotice"/>
+
+### InvalidEmailNotice
+
+Value of field with type `email` is not valid. 
+
+#### References:
+* [Field Types Description](http://gtfs.org/reference/static/#field-types)
+* [Apache Commons EmailValidator](https://commons.apache.org/proper/commons-validator/apidocs/org/apache/commons/validator/routines/EmailValidator.html)
```

Comment:
The above description could use an explanation of why the Apache Commons EmailValidator is listed here. I think there is one in the notice/validator? Maybe just copy/paste here?

Same with UrlValidator, and anything else that references outside libraries.